packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/RestClientProvider.cs

```diff
@@ -186,26 +145,50 @@ private IReadOnlyList<MethodBodyStatement> GetSetContent(HttpRequestApi request,
             return contentParam is null ? [] : [request.Content().Assign(contentParam).Terminate()];
         }
 
-        private PropertyProvider GetClassifier(InputOperation operation)
+        private Dictionary<string, PropertyProvider> BuildPipelineMessage20xClassifiers()
         {
-            if (operation.HttpMethod == HttpMethod.Head.ToString())
-                return _classifier2xxAnd4xxProperty;
+            // Contains a mapping of classifier status code suffixes to their corresponding pipeline message classifier property
+            Dictionary<string, PropertyProvider> classifiers = [];
 
-            var response = operation.Responses.First(r => !r.IsErrorResponse); //should only be one of these
-
-            if (response.StatusCodes.Count == 1)
+            foreach (var inputOperation in _inputClient.Operations)
             {
-                return response.StatusCodes[0] switch
+                if (TryGetPipelineMessageClassifierSuffix(inputOperation, out var classifierNameSuffix) &&
+                    !classifiers.ContainsKey(classifierNameSuffix))
                 {
-                    200 => _classifier200Property,
-                    201 => _classifier201Property,
-                    202 => _classifier202Property,
-                    204 => _classifier204Property,
-                    _ => throw new InvalidOperationException($"Unexpected status code {response.StatusCodes[0]}")
-                };
+                    var classifierBackingField = new FieldProvider(
+                        FieldModifiers.Private | FieldModifiers.Static,
+                        ClientModelPlugin.Instance.TypeFactory.StatusCodeClassifierApi.ResponseClassifierType,
+                        $"_pipelineMessageClassifier{classifierNameSuffix}",
+                        this);
+
+                    var classifierProperty = new PropertyProvider(
+                        null,
+                        MethodSignatureModifiers.Private | MethodSignatureModifiers.Static,
+                        ClientModelPlugin.Instance.TypeFactory.StatusCodeClassifierApi.ResponseClassifierType,
+                        classifierBackingField.Name.Substring(1).ToCleanName(),
+                        new ExpressionPropertyBody(
+                            classifierBackingField.Assign(This.ToApi<StatusCodeClassifierApi>().Create(GetSuccessStatusCodes(inputOperation)))),
+                        this)
+                    {
+                        BackingField = classifierBackingField
+                    };
+
+                    classifiers[classifierNameSuffix] = classifierProperty;
+                }
+            }
+
+            return classifiers;
+        }
+
+        private PropertyProvider GetClassifier(InputOperation operation)
+        {
+            if (TryGetPipelineMessageClassifierSuffix(operation, out var classifierSuffix) &&
```

Comment:
Would it make sense to have a dictionary of Operation to PropertyProvider rather than string to PropertyProvider? Then we wouldn't need to do the computation to get the suffix here.

Reply:
Hmm, I think if we have the dictionary be `<InputOperation, PropertyProvider>` there might be additional work to ensure that any operations that require the same classifier map to the same property. When we construct the fields and properties for the provider, we'll also need to check for duplicate classifier properties. Though if I'm missing something, or we think the suffix approach is not the most performant I'm open to more suggestions.

Reply:
Discussed offline and we went with constructing a dictionary of `List<int>` to `PropertyProvider`.